sentinel-core/src/main/java/com/alibaba/csp/sentinel/slots/block/degrade/circuitbreaker/AbstractCircuitBreaker.java

```diff
@@ -99,11 +103,24 @@ protected boolean fromCloseToOpen(double snapshotValue) {
         return false;
     }
 
-    protected boolean fromOpenToHalfOpen() {
+    protected boolean fromOpenToHalfOpen(Context context) {
         if (currentState.compareAndSet(State.OPEN, State.HALF_OPEN)) {
             for (CircuitBreakerStateChangeObserver observer : observerRegistry.getStateChangeObservers()) {
                 observer.onStateChange(State.OPEN, State.HALF_OPEN, rule, null);
             }
+            Entry entry = context.getCurEntry();
+            entry.whenComplete(new BiConsumer<Context, Entry>() {
```

Comment:
1. Thread-safe for`whenComplete`.
2. Other requests blocked by half-open state also can transform half-open to open, so the one request allowed by half-state maybe not  transform half-open to close.

Reply:
> * Thread-safe for`whenComplete`.
> * Other requests blocked by half-open state also can transform half-open to open, so the one request allowed by half-state maybe not  transform half-open to close.

Please refer to the workflow chart above.
I think the design of `Entry` is non-thread-safe because it stores information of single request. One request cannot be paralleled, right?

Talk to this piece of code the mutex between entries(requests) is ensured by CAS operation (from OPEN to HALF-OPEN). So only one request can perform `open -> half -> open` or `open -> half -> closed`.

Or is there any other scenario I missed?

Reply:
my mistake, no problem.

Reply:
😀